### ISSUE TITLE
voice helpers

### ISSUE THREAD

**Issue by ianjuma** [AfricasTalkingLtd/africastalking-node.js#15]
the voice helpers are very interesting; what approach do we take? plain xml response? or command to construct xml ? The helpers will be great, this will make voice so easy :+1: 

``` javascript
voice.say("X is a letter! It's not a word") 
```

Here the approach is to wrap `say` around root xml response - what of nested commands? and attributes, how to create the xml response? 

Of course this is for incoming calls, with the callback url handler. 
`voice.call` is ok. I don't think there's much there. I mean initiating a call.

``` javascript
voice.say("say something", voice="woman") // ?
```

`Say` then `Play` a URL

``` node
/*
<Say>Please listen to our awesome record</Say>
<Play url="http://www.myvoicemailserver.com/audio/vmail.wav"/>
*/
voice.say("Hi, listen to", play=url)
```

and how do we wrap `getDigits` when we `say`?

``` node
voice.say("say something I'm giving ... ", voice=woman, getDigits=True, record=True) // ?
// sadly without named parameters
// also, recording - maybe an attribute to be set
```

call redirect?

``` node
if (dtmfDigit === '9') {
   voice.redirect('http://some_url_callback');
}
```

then, this would be expected.

``` xml
<Response>
  <Dial phoneNumbers="+254711XXXYYY,+254733YYYZZZ,test@ke.sip.africastalking.com" ringBackTone="http://mymediafile.com/playme.mp3" record="true" sequential="true"/>
</Response>
```

maps to

``` node
// callerId specific to calling out from SIP
voice.dial(phoneNumbers=phones, maxDurationInSec=10, record=True, 
                ringBackTone="some_url", sequential="True", callerId="some_num");
```

Problem with this is that we need all possible mappings, and I think there's many cases, we might miss a few and not generate a proper functional XML string, but the XML response will still work.

So, basically we generate this by passing attributes on a command, is this a good approach?
we also need a SIP section, will look at this later.

**Comment by ianjuma** [comment]
also, ES6 does not support named parameters :-1:  - so, maybe pass an options object instead. and have this as the only argument, to just have this as the only named parameter object.

**Comment by pbombo** [comment]
Passing an options object seems like a decent compromise and it could still have default values something like below is still possible

```javascript
function (options) {
    var someSetting = options.someSetting || defaultValue;
}
```

**Comment by pbombo** [comment]
[This](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Functions/Default_parameters) is also looking interesting, default parameters in ES6

**Comment by ianjuma** [comment]
I think my reaction was "JS can do that?" lol, I assumed it would, it's very pythonic. Better than ``` ||```

**Comment by aksalj** [comment]
I think it can be done well even without named parameters. Something like this would be nice:

```javascript
let str = "Salut!", anotherStr, url;
let options = { // action's secondary attributes
   voice: 'dude',
   record: false
   //...
};
voice.say(str, options)
     .play(url)
     .then(success)
     .catch(error);

voice.getDigits(options)
     .say(str, options)
     .then(success)
     .catch(error);
```

But quick question: Is the documentation at [http://docs.africastalking.com/voice](http://docs.africastalking.com/voice) up to date? I remember @ianjuma you said there were some big changes coming to voice... 😕

**Comment by pbombo** [comment]
@aksalj I think the idea was to avoid user having to do this

```javascript
let options = { // action's secondary attributes
   voice: 'dude',
   record: false
   //...
};
```
that is the real value of default parameters IMO.

Ideally you'd just have ```javascript voice.say(str)``` without the need to set options

**Comment by aksalj** [comment]
@ianjuma is the problem **named parameters** or **default parameters**?

With named params (which is what I think @ianjuma was referring to) you can have both this:
`voice.say(str, voice='Female', record=false)` 
and this: 
`voice.say(str, record=false, voice='Female')` 
be the exact same call. 
That way you can avoid passing params in a given order (therefore passing only those you need). JavaScript does not support named parameters yet afaik. But we can "simulate" this behavior with objects like this:
```javascript
let options = { // only set the options you need, otherwise default values will be used
   voice: 'Female',
   record: false
   //...
};
voice.say(str, options);
// if you don't need to override the default options, just do
voice.say(str); 
```  
Then `say()` is implemented like @pbombo said with `something = options.something || defaultValue`.

**Comment by ianjuma** [comment]
@pbombo I wish we didn't have to do that; I think the order of params should be free from the user, so
 we might have the ```options``` set. My problem is with attributes, it just seems un-conventional to have those in ```options```. Python problems :( but that's the way to go @aksalj I agree.

@aksalj some of the commands were updated, and there parameters too. ```Say, Play, Record, Redirect, Reject``` should be ok now. I've been slow with that, sadly :disappointed: Users on the API have been creating ```XML``` responses, I really want this helper approach, I think it will make working with Voice easy. 
Also @aksalj I can map a number in your account, for testing with this?

now, order of commands? How are we making this response. Responses are usually simple and straight forward. So, are we grabbing the response string and returning it?

```node
voice.say(str, options)
     .play(url)
     .then(success)
     .catch(error);
```

**Comment by habbes** [comment]
May I contribute to this discussion. I also go for having an options object. Considering the lack of default params and strong intellisense-like features, one ends up looking up the docs a lot just to remember the order of params.

**Comment by ianjuma** [comment]
@habbes That's the way to go now; @aksalj had already highlighted that.

**Comment by aksalj** [comment]
Could something like [this](https://github.com/aksalj/africastalking-node.js/tree/voice-call-handler#handle-call) be a good solution?

```javascript
// example (express)
app.post('/callback/voice', new AfricasTalking.VOICE.CallHandler((phoneCall) => {
    
    var session = phoneCall.session;
    //...
    
    if (phoneCall.isActive) {
        //...
    } else if (phoneCall.isCompleted){
        //...
    }
    
    
    var response = voice.builder()
        .say("UI", {voice: "woman"})
        .play("DEd")
        .getDigits("SW", {say:{voice: "man"}})
        //... more say(), play(), record(), etc.
        .build();
        
    phoneCall.respond(null, response);
   
    
}));
```

**Comment by ianjuma** [comment]
that's perfect! awesome! this will make it so easy to build voice apps on node :100:

**Comment by danmbeyah** [comment]
@ianjuma is the voice library updated with `CallHandler()`?
